crates/utils/src/settings.rs

```diff
@@ -1,27 +1,53 @@
-use crate::location_info;
+use crate::{location_info, LemmyError};
 use anyhow::Context;
-use config::{Config, ConfigError, Environment, File};
+use deser_hjson::from_str;
+use merge::Merge;
 use serde::Deserialize;
-use std::{env, fs, io::Error, net::IpAddr, sync::RwLock};
+use std::{
+  env,
+  fs,
+  io::Error,
+  net::{IpAddr, Ipv4Addr},
+  sync::RwLock,
+};
 
-static CONFIG_FILE_DEFAULTS: &str = "config/defaults.hjson";
 static CONFIG_FILE: &str = "config/config.hjson";
 
-#[derive(Debug, Deserialize, Clone)]
+#[derive(Debug, Deserialize, Clone, Merge)]
 pub struct Settings {
   pub setup: Option<Setup>,
-  pub database: DatabaseConfig,
-  pub hostname: String,
-  pub bind: IpAddr,
-  pub port: u16,
-  pub tls_enabled: bool,
-  pub jwt_secret: String,
-  pub pictrs_url: String,
-  pub iframely_url: String,
-  pub rate_limit: RateLimitConfig,
+  pub database: Option<DatabaseConfig>,
+  pub hostname: Option<String>,
+  pub bind: Option<IpAddr>,
+  pub port: Option<u16>,
+  pub tls_enabled: Option<bool>,
+  pub jwt_secret: Option<String>,
+  pub pictrs_url: Option<String>,
+  pub iframely_url: Option<String>,
+  pub rate_limit: Option<RateLimitConfig>,
   pub email: Option<EmailConfig>,
-  pub federation: FederationConfig,
-  pub captcha: CaptchaConfig,
+  pub federation: Option<FederationConfig>,
+  pub captcha: Option<CaptchaConfig>,
+}
+
+impl Default for Settings {
+  fn default() -> Self {
+    Self {
+      database: Some(DatabaseConfig::default()),
+      rate_limit: Some(RateLimitConfig::default()),
+      federation: Some(FederationConfig::default()),
+      captcha: Some(CaptchaConfig::default()),
+      email: None,
+      setup: None,
+      hostname: None,
+      bind: Some(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0))),
+      port: Some(8536),
+      tls_enabled: Some(true),
+      jwt_secret: Some("changeme".into()),
+      pictrs_url: Some("http://pictrs:8080".into()),
+      iframely_url: Some("http://iframely".into()),
+    }
+  }
```

Comment:
The defaults are now implemented in code, meaning the `defaults.hjson` is entirely unused. (But its still useful to have as it shows an expanded and commented version of the config.)

Reply:
I think you should split settings into multiple files, cause its already pretty long. And then you can put the default impls all in the same file, so they are right next to each other. Plus another file for the structs, and one (or more) for all impls.

Reply:
That's really not too bad, its only ~ 300 lines, that's probably less than most our files. IMO with rust you could have everything in one file, what really matters is proper scoping, splitting specific tasks / concerns into modules, and being able to trace down and find references easily.

Reply:
Then the majority of our files are too long. Splitting them up will make proper scoping much easier, and will make it easier to find the code you are looking for.